Port shared_ptr optimization in std::shared_ptr to intrusive_ptr

mcfi opened the pull request:
Summary:
Please see D21021645 for details about the optimization and why it's beneficial.

A similar change has been added to libstdc++ as well, see https://github.com/gcc-mirror/gcc/commit/dbf8bd3c2f2cd2d27ca4f0fe379bd9490273c6d7

Rollback Plan:

Reviewed By: yfeldblum

Differential Revision: D81960754

yfeldblum commented:
@swolchok I like it. That would remove the need for undefined behavior. They might not be able to do it in libstdc++ because it's not abi-compatible and that's something they specifically care about, but it should be fine in pytorch.

mcfi committed:
Port shared_ptr optimization in D21021645 to intrusive_ptr (#162784)

Summary:

Please see D21021645 for details about the optimization and why it's beneficial.

A similar change has been added to libstdc++ as well, see https://github.com/gcc-mirror/gcc/commit/dbf8bd3c2f2cd2d27ca4f0fe379bd9490273c6d7

Test Plan:
SL runs below showed that the diff consistently reduced the total CPU time by ~0.1%.

https://www.internalfb.com/servicelab/experiment/5902177629/

https://www.internalfb.com/servicelab/experiment/6500025398/

V10:

Arm64:
https://www.internalfb.com/servicelab/experiment/6100105254/
https://www.internalfb.com/servicelab/experiment/6300105717/

x64:
https://www.internalfb.com/servicelab/experiment/6100105261/
https://www.internalfb.com/servicelab/experiment/6200105148/

Rollback Plan:

Reviewed By: yfeldblum

Differential Revision: D81960754

mcfi commented:
> @mcfi it does occur to me that the existing code assumes refcounts never overflow a 32-bit int anyway, so we might be able to do this with well-defined behavior by fusing the refcounts into a single std::atomic<uint64_t> for real. The code change is a bit more invasive (you have to define an interface to the fused thing that wraps a std::atomic<uint64_t> on the platforms you want to affect in this diff and wraps the existing pair of 32-bit ints otherwise), but should be well-defined. (since neither count overflows, we can just increment/decrement weakcount by adding/subtracting 0x100000000UL.) What do you think?

> @swolchok I like it. That would remove the need for undefined behavior. They might not be able to do it in libstdc++ because it's not abi-compatible and that's something they specifically care about, but it should be fine in pytorch.

It's indeed a good idea that can remove undefined behavior and simplify the code. However, my concern is that adding/subtracting 0x100000000UL cannot be done in one instruction, so performance of weak pointers may downgrade. Moreover, if the two counters are merged in one, contention might be higher than before, which downgrades performance as well. Is this task something good for the open-source community to pick up? :)

Skylion007 commented on c10/util/intrusive_ptr.h:
```diff
@@ -283,23 +283,55 @@ class intrusive_ptr final {
+          std::atomic<uint32_t>::is_always_lock_free &&
+          sizeof(std::atomic<uint64_t>) == 8 &&
+          sizeof(std::atomic<uint32_t>) == 4) {
+        auto both_counts_ =
+            reinterpret_cast<std::atomic<uint64_t>*>(&target_->refcount_);
```
Hmm c10::bitcast won't make this defined either, right?

Skylion007 commented:
> > @mcfi it does occur to me that the existing code assumes refcounts never overflow a 32-bit int anyway, so we might be able to do this with well-defined behavior by fusing the refcounts into a single std::atomic<uint64_t> for real. The code change is a bit more invasive (you have to define an interface to the fused thing that wraps a std::atomic<uint64_t> on the platforms you want to affect in this diff and wraps the existing pair of 32-bit ints otherwise), but should be well-defined. (since neither count overflows, we can just increment/decrement weakcount by adding/subtracting 0x100000000UL.) What do you think?
> 
> 
> 
> > @swolchok I like it. That would remove the need for undefined behavior. They might not be able to do it in libstdc++ because it's not abi-compatible and that's something they specifically care about, but it should be fine in pytorch.
> 
> 
> 
> It's indeed a good idea that can remove undefined behavior and simplify the code. However, my concern is that adding/subtracting 0x100000000UL cannot be done in one instruction, so performance of weak pointers may downgrade. Moreover, if the two counters are merged in one, contention might be higher than before, which downgrades performance as well. Is this task something good for the open-source community to pick up? :)

Does godbolt show that compilers cannot fuse this into one instruction?

mcfi commented:
> @mcfi The cost is likely negligible. Worth double checking, of course. But the costs we care about here are primarily interlocked instructions, barrier instructions, seq-cst, and unpredictable branch conditions or jump targets. Changing the representation would affect none of these.
> 
> Incrementing by 1ull<<32 would probably just be a larger-encoded instructions (x86) or an extra math instruction (arm). Would probably be lost in the pipeline. (Still worth checking. Because avoiding UB is a worthy goal.)

Sounds good, I’ll explore that next week.
